README.md
[![Version](https://shields.io/badge/VERSION-1.5.2-blue?style=for-the-badge)](https://github.com/ghasemdev/affogato/releases/tag/1.5.2)
[![Kotlin](https://shields.io/badge/Kotlin-1.7.10-pink?style=for-the-badge)](https://kotlinlang.org/)
[![API](https://shields.io/badge/Api-+21-green?style=for-the-badge)](https://www.android.com/)
[![License MIT](https://shields.io/badge/LICENSE-MIT-orange?style=for-the-badge)](https://opensource.org/licenses/MIT)

![affogato_img](assets/affogato.png)

# Affogato

Affogato is a multipurpose library for Jetpack Compose, Android and Kotlin.

### [Core-ktx](https://github.com/ghasemdev/affogato/wiki/Core-Ktx)

Core-ktx is a Kotlin library that provides a set of Kotlin extensions. Like collections, operations,
string, math, random, date time, color, block scope and try-catch functions. Also, `TimeAgo` is my
favorite extension for times 😄.

```kotlin
TimeAgo.format(now()) // a moment ago
now().timeAgo(locale = TimeAgo.Locale.EN) // a moment ago
```

### [Coroutines](https://github.com/ghasemdev/affogato/wiki/Coroutines)

Coroutines module contains suspended try-catch from Core-ktx and `DispatchersProvider` for handling
coroutines context in main and test source.

```Kotlin
// Normal way
suspend fun foo() = withContext(Dispatchers.Default) {
  return try {
    // do something return value
  } catch (_: Exception) {
    null
  }
}

// Affogato way
suspend fun foo() = suspendedTryCatchNull {
  // do something return value
}
```

### [Logger](https://github.com/ghasemdev/affogato/wiki/Logger)

Logger module contains a simple logger for Android. It's a wrapper for `Log` and`Logger` with some
extensions.

```Kotlin
// Setup
Logger.tag = "MyApp"

Logger.isDebug = BuildConfig.DEBUG // if debug mode false onRelease will be called.
Logger.onRealease = { tag, message, throwable ->
  // do something
}

// Extension
logcat { "Hello Affogato" }
logcat(tag = "TAG") { "Hello Affogato" }
logcat(LogLevel.INFO) { "Hello Affogato" }
try {
  // do something
} catch (e: Exception) {
  logcat(throwable = e) { "Hello Affogato" }
}
```

### [Metrica-ktx](https://github.com/ghasemdev/affogato/wiki/Metrica-Ktx)

This module adds Yandex Metrica to your project. You can use it for crashlytics and reporting events
and push notifications.

```kotlin
yandexMetrica("api-key") {
  // configuration
}

@Serializable
data class Foo(val bar: String)
YandexMetricaX.reportEvent("event_name", Foo("bar"))
```

### [Okhttp](https://github.com/ghasemdev/affogato/wiki/Okhttp)

Okhttp builder is part of this module. You can use it for building okhttp client with a simple
syntax. Also, you can track download and upload progress with `DownloadStreamResponseBody`
and `UploadStreamRequestBody`.

```kotlin
val client = okHttp {
  addInterceptor(HttpLoggingInterceptor())
  cache(Cache(cacheDir, cacheSize))
}
```

### [Structure](https://github.com/ghasemdev/affogato/wiki/Structure)

This module contains useful structure class like `DataState` and `EntityMapper`. Also, it contains
`SingletonHolder` for creating singleton classes or `ObjectPool` for creating heavy objects to save
time. You can use checker class for validate phone, email and password.

```kotlin
class SharedPref private constructor(private val context: Context) {
  // ...
  companion object : SingletonHolder<SharedPref, Context>(::SharedPref)
}

StringPool["key"] = "token"
StringPool["key"] // token

Phone("09123456789").isValid() // true
Phone("09123456789").format("IR") // +98 912 345 6789
```

### [Compose Unit Size](https://github.com/ghasemdev/affogato/wiki/Unit-Size)

The unit is one of the **Jetpack Compose** modules of this library, which can support relative
screen sizes in `sdp`, `ssp` and custom create with `@Dimen`. Also, we can
use `rememberWindowSize()` to know in which device we are (Compact, Medium, Expanded) or
use `postureState` to build adaptive and responsive UIs in **Foldables**.

```kotlin
@Dimen(type = "dp", values = ["320:70", "480:80", "600:180", "720:180"])
val icon = 80.dp

dimen.icon // 80.dp in device with width 480.dp
```

# Links

Affogato is currently available for maven/gradle builds
at [Jitpack](https://jitpack.io/#ghasemdev/affogato/1.5.2) (
read [Getting started](https://github.com/ghasemdev/affogato/wiki/Getting-Started)).

* [Wiki](https://github.com/ghasemdev/affogato/wiki)
* [Change log](https://github.com/ghasemdev/affogato/blob/master/CHANGELOG.md)

# Download

### Gradle Setup

Step 1. Add the JitPack repository to your `build.gradle` file

```
allprojects {
    repositories {
        ...
        maven { url 'https://jitpack.io' }
    }
}
```

Step 2. Add the dependency

```
dependencies {
    implementation 'com.github.ghasemdev.affogato:AFFOGATO_MODULE:AFFOGATO_VERSION'
}
```

---

### Maven Setup

Step 1. Add the JitPack repository to your `build.gradle` file

```
<repositories>
    <repository>
        <id>jitpack.io</id>
        <url>https://jitpack.io</url>
    </repository>
</repositories>
```

Step 2. Add the dependency

```
<dependency>
    <groupId>com.github.ghasemdev.affogato</groupId>
    <artifactId>AFFOGATO_MODULE</artifactId>
    <version>AFFOGATO_VERSION</version>
</dependency>
```

---

### License

```
The MIT License (MIT)

Copyright (c) 2022 ghasemdev

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and
associated documentation files (the "Software"), to deal in the Software without restriction,
including without limitation the rights to use, copy, modify, merge, publish, distribute,
sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial
portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT
NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES
OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
```


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).

## [🐛 v1.5.2 ] - 2022-08-24

### Fix affogato-okhttp-android gradle issue

## [🐛 v1.5.1 ] - 2022-08-23

### Fix min sdk issue and some improvements

## [🎉 1.5.0 Structure, Logger, OkHttp] - 2022-08-22

### 🚀 Features

* Add `okhttp-android` module
	* okhttp builder extension `okhttp {}` , `Response.new {}` and `OkHttpClient.new {}`.
	* download and upload progress listener `DownloadStreamResponseBody`, `UploadStreamRequestBody`.
* Add `logger-android` module
	* `LogLevel` enum.
	* `logcat` extension.
	* Singleton `Logger`.

### 💥 Breaking Changes

* Rename checker class from **structure** module.

## [🎉 1.4.0 Structure, Metrica] - 2022-08-21

### 🚀 Features

* Add New Structure Class.
	* lazy singleton object
	* object pool pattern
	* validator class `EmailChecker`, `PasswordChecker`, `PhoneChecker` and `IRPhoneChecker`
* Add `Metrica` Module.
	* metrica builder `yandexMetrica("api-key") {}`, `yandexMetricaConfig("api-key") {}`
	  , `withLogs(isDebugMode)`, `withSessionTimeout(duration)`
	* metrica reporter for serialized class `YandexMetricaX.reportEvent("key", data-class)`
	* metrica profile builder `metricaProfile {}`, `withValue(Int)`

### 💥 Breaking Changes

* Remove additional classes from affogato-coroutines-android

### 💭 Other

* rename `block` -> `runBlock`
* rename `isNotNullOrEmpty` -> `isNotNullNotEmpty`, `isNotNullOrBlank` -> `isNotNullNotBlank`

## [🎉 1.3.0 Structure] - 2022-08-18

### 🚀 Features

* Add local const val to `TimeAgo` class instead of used invalid locale strings.
* New structure module with below features:
	* `EntityMapper` - Convert entity to domain model and vice versa.
	* `DataState` - Contains three different state -> Loading, Error and Success.

## [🎉 1.2.0 Coroutines] - 2022-08-16

### 🚀 Features

* `DispatchersProvider`.
* suspended try-catch.

## [🐛 v1.1.2 Core-ktx] - 2022-08-15

### 🚀 Features

* add weeks duration (`2.weeks`)

### 🐛 Fixes

* Fix rounding issue for TimeAgo.kt

## [🐛 v1.1.1 Core-ktx] - 2022-08-14

### 💥 Breaking Changes

* Rename date parameter in LookupMessages to dateInMillis
* Rename toString -> format in all datetime extensions

### 🐛 Fixes

* Fix duplicated $ sign in `ar` and `fa` TimeAgo Messages

## [🎉 v1.1.0 Core-ktx] - 2022-08-13

### 🚀 Features

* Collections Function (built-in and get functions based on Range and Progression, sorting map,
  rotate, head, tail and ...)
* Operations Function (collection operation add, sub, mul, div and rem, string times and some
  boolean operator)
* String Function (formatting and get functions based on Range and Progression, separate,
  removeDecimalPart, rotate and ... )
* Math Function (math for collection like median, std, variance and more, BD and BI, round with
  digit, infix pow and degree to radian, ... )
* Random Function (choice, choices and nextInt-long based on Progression)
* Date Time Function (convert all type of date time to each other :), now functions and time ago)
* Color Converter (rgb to hex, hex to rgb)
* Block Scope Function
* Try-Catch Functions and Counter

### 💥 Breaking Changes

* Change time ago messages package
* Remove onDate function from LookupMessages class
* Change `as...` function -> `to...` function for date time converters

## [🎉 v1.1.0-beta02 Core-ktx] - 2022-08-13

### 🚀 Features

* Add factory pattern to set local in `TimeAgo`
* Add new languages for `TimeAgo`
* Add cut off, local number, week support for `TimeAgo`
* Add new compact format function for numbers

### 💥 Breaking Changes

* Change time ago messages package
* Remove onDate function from LookupMessages class
* Change `as...` function -> `to...` function for date time converters

### 🐛 Fixes

* Fix TimeAgo arabic issue

### 🧪 Tests

* Add more tests

## [🎉 v1.1.0-beta01 Core-ktx] - 2022-08-10

### 🚀 Features

* Collections Function (built-in and get functions based on Range and Progression, sorting map,
  rotate, head, tail and ...)
* Operations Function (collection operation add, sub, mul, div and rem, string times and some
  boolean operator)
* String Function (formatting and get functions based on Range and Progression, separate,
  removeDecimalPart, rotate and ... )
* Math Function (math for collection like median, std, variance and more, BD and BI, round with
  digit, infix pow and degree to radian, ... )
* Random Function (choice, choices and nextInt-long based on Progression)
* Date Time Function (convert all type of date time to each other :), now functions and time ago)
* Color Converter (rgb to hex, hex to rgb)
* Block Scope Function
* Try-Catch Functions and Counter

## [🎉 v1.0.0 Affogato Unit] - 2022-07-27

### 🚀 Features

* sdp and ssp unit
* Dimen tag for create auto dimensions
* Window Size, Global Window Size, rememberWindowSize()
* Window Orientation, isLandscape
* Posture, postureState, postureFlow, ...
* Hinge Size, hingeWidthDp, hingeHeightDp, ...
* WindowLayoutInfo, windowLayoutInfoState, windowLayoutInfo, ...

## [🎉 v1.0.0-beta05 Affogato Unit] - 2022-07-26

### 🐛 Fixes

* Fix orientation in landscape

## [🎉 v1.0.0-beta04 Affogato Unit] - 2022-07-17

### 🐛 Fixes

* Fix Build Issue

## [🎉 v1.0.0-beta03 Affogato Unit] - 2022-07-17

### 🚀 Features

* Add Window Size
* Add rememberWindowSize()

## [🎉 v1.0.0-beta02 Affogato Unit] - 2022-07-16

### 🚀 Features

* Add new sample

### 🧪 Tests

* Add tests for dimen tag validator

### 💭 Other

* Add doc and update readme

## [🎉 v1.0.0-beta01 Affogato Unit] - 2022-07-15

### 🚀 Features

* Add sdp and ssp unit
* Add Dimen tag for create auto dimensions


assets/DimenProvider.kt
package com.parsuomash.affogato

import androidx.compose.runtime.Composable
import androidx.compose.runtime.CompositionLocalProvider
import androidx.compose.runtime.remember
import androidx.compose.runtime.staticCompositionLocalOf
import androidx.compose.ui.platform.LocalConfiguration
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.TextUnit
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp

/**
 * Provide dimens for multiscreen support.
 * in Theme.kt add this function
 * ```Kotlin
 * ProvideDimens {
 *   MaterialTheme(
 *     colors = colors,
 *     typography = Typography,
 *     shapes = Shapes,
 *     content = content
 *   )
 * }
 * ```
 *
 * @param dimensions
 * @param content
 */
@Composable
public fun ProvideDimens(
    dimensions: Dimensions = defaultDimens(),
    content: @Composable () -> Unit
) {
    val dimensionSet = remember { dimensions }
    CompositionLocalProvider(LocalAppDimens provides dimensionSet, content = content)
}

private val LocalAppDimens = staticCompositionLocalOf {
    Dimensions()
}

public val dimen: Dimensions
    @Composable
    get() = LocalAppDimens.current

@Composable
private fun defaultDimens(): Dimensions {
    val configuration = LocalConfiguration.current
    val screenWidth = configuration.screenWidthDp

    return when {
        screenWidth < 320 -> sw320Dimensions
        screenWidth in 320..480 -> sw320Dimensions
        screenWidth in 480..600 -> sw480Dimensions
        screenWidth in 600..720 -> sw600Dimensions
        else -> sw720Dimensions
    }
}

public data class Dimensions(
     val icon: Dp = com.parsuomash.affogato.icon,
     val padding: Dp = com.parsuomash.affogato.padding,
     val space: Dp = com.parsuomash.affogato.space,
     val heightButton: Dp = com.parsuomash.affogato.heightButton,
     val fontSize: TextUnit = com.parsuomash.affogato.fontSize,
)

private val sw320Dimensions = Dimensions(
     icon = 70.dp,
     padding = 14.dp,
     space = 26.dp,
     heightButton = 50.dp,
     fontSize = 14.sp,
)

private val sw480Dimensions = Dimensions(
     icon = 80.dp,
     padding = 16.dp,
     space = 32.dp,
     heightButton = 50.dp,
     fontSize = 20.sp,
)

private val sw600Dimensions = Dimensions(
     icon = 180.dp,
     padding = 18.dp,
     space = 34.dp,
     heightButton = 70.dp,
     fontSize = 30.sp,
)

private val sw720Dimensions = Dimensions(
     icon = 180.dp,
     padding = 20.dp,
     space = 36.dp,
     heightButton = 80.dp,
     fontSize = 32.sp,
)
